src/engine/MaterializedViews.cpp

```diff
@@ -0,0 +1,504 @@
+#include "util/MemorySize/MemorySize.h"
+#include "util/ProgressBar.h"
+#include "util/Views.h"
+
+// _____________________________________________________________________________
+MaterializedViewWriter::MaterializedViewWriter(
+    std::string name, const qlever::Qlever::QueryPlan& queryPlan,
+    ad_utility::MemorySize memoryLimit,
+    ad_utility::AllocatorWithLimit<Id> allocator)
+    : name_{std::move(name)},
+      memoryLimit_{std::move(memoryLimit)},
+      allocator_{std::move(allocator)} {
+  MaterializedView::throwIfInvalidName(name_);
+  auto [qet, qec, parsedQuery] = queryPlan;
+  AD_CORRECTNESS_CHECK(qet != nullptr);
+  AD_CORRECTNESS_CHECK(qec != nullptr);
+  qet_ = qet;
+  qec_ = qec;
+  parsedQuery_ = std::move(parsedQuery);
+}
+
+// _____________________________________________________________________________
+void MaterializedViewWriter::writeViewToDisk(
+    std::string name, const qlever::Qlever::QueryPlan& queryPlan,
+    ad_utility::MemorySize memoryLimit,
+    ad_utility::AllocatorWithLimit<Id> allocator) {
+  MaterializedViewWriter writer{std::move(name), queryPlan,
+                                std::move(memoryLimit), std::move(allocator)};
+  writer.computeResultAndWritePermutation();
+}
+
+// _____________________________________________________________________________
+std::string MaterializedView::getFilenameBase(std::string_view onDiskBase,
+                                              std::string_view name) {
+  return absl::StrCat(onDiskBase, ".view.", name);
+}
+
+// _____________________________________________________________________________
+std::string MaterializedViewWriter::getFilenameBase() const {
+  return MaterializedView::getFilenameBase(qec_->getIndex().getOnDiskBase(),
+                                           name_);
+}
+
+// _____________________________________________________________________________
+MaterializedViewWriter::ColumnNamesAndPermutation
+MaterializedViewWriter::getIdTableColumnNamesAndPermutation() const {
+  AD_CONTRACT_CHECK(
+      parsedQuery_.hasSelectClause(),
+      "Materialized views may only be built from 'SELECT' statements. "
+      "'CONSTRUCT', 'ASK' and update queries are not allowed.");
+
+  auto targetVarsAndCols =
+      qet_->selectedVariablesToColumnIndices(parsedQuery_.selectClause());
+  auto numCols = targetVarsAndCols.size();
+  AD_CONTRACT_CHECK(numCols >= 4,
+                    "Currently the query used to write a materialized view "
+                    "needs to have at least four columns.");
+
+  std::vector<std::string> targetVars;
+  targetVars.reserve(numCols);
+  std::vector<ColumnIndex> columnPermutation;
+  columnPermutation.reserve(numCols);
+
+  for (const auto& varAndColIndex : targetVarsAndCols) {
+    AD_CONTRACT_CHECK(varAndColIndex.has_value());
+    targetVars.push_back(varAndColIndex.value().variable_);
+    columnPermutation.push_back(varAndColIndex.value().columnIndex_);
+  }
+  return {targetVars, columnPermutation};
```

Comment:
I don't quite understand why you don't simple store the original format, and have to do sthe split up:)